Add Map.split_with/2 and Keyword.split_with/2

camilleryr opened the pull request:
As per [this discussion](https://groups.google.com/g/elixir-lang-core/c/2OEFTGEHPMM) in mailing group - here is a pr that adds implementations of `split_with/2` to the Map and Keyword modules

v0idpwn commented on lib/elixir/lib/keyword.ex:
```diff
@@ -1166,6 +1166,49 @@ defmodule Keyword do
+  `fun`.
+
+  Splits the given `keywords` into two keyword lists by calling the provided
+  `fun` witch receives each `{key, value}` pair in the `keywords` as its only
```
```suggestion
  `fun` which receives each `{key, value}` pair in the `keywords` as its only
```

camilleryr replied:
Whoops - sorry! Fixed!

v0idpwn commented on lib/elixir/lib/keyword.ex:
```diff
@@ -1166,6 +1166,49 @@ defmodule Keyword do
+    end
+
+    {while_true, while_false} = :lists.foldl(fun, {[], []}, keywords)
+    {:lists.reverse(while_true), :lists.reverse(while_false)}
```
As there are no guarantees regarding keyword list ordering, I think this reverse isn't necessary.

camilleryr replied:
While its true that the ordering is not guaranteed - many operations, including `split/2` do retain the ordering, so I would be worried that the lack of consistency would be more troublesome than the actual change to the order of the elements.

I would be happy to drop the reverse, just wanted to make note of the reasoning that I included it!

josevalim replied:
Correct! We can just return the result of foldl!

josevalim replied:
@camilleryr we were writing at the same time. Actually, you are right, we need to reverse it, otherwise we affect the result of operations like `Keyword.get_values`. So please do keep the reverse!

palexanderm replied:
@josevalim Couldn't the reverse be omitted if `:lists.foldr` was used instead? Is there a performance aspect which makes `:lists.foldl` + reversing better than just `:lists.foldr`?

josevalim replied:
Good question. foldr is just doing a reverse before folding, so at the end the result is the same.  :)

camilleryr replied:
Since they are the same - Ill swap it over to `foldr` so I can drop the extra line of code.  I also assume doing a single reverse might be more efficient than doing two reverses after the list has been split?

josevalim commented on lib/elixir/lib/map.ex:
```diff
@@ -838,6 +838,55 @@ defmodule Map do
+  a tuple with the first map containing all the elements in `map` for which
+  applying `fun` returned a truthy value, and a second map with all the elements
+  for which applying `fun` returned a falsy value (`false` or `nil`).
+
+  See the examples below.
```
```suggestion
```

josevalim commented on lib/elixir/lib/keyword.ex:
```diff
@@ -1166,6 +1166,49 @@ defmodule Keyword do
+  elements in `keywords` for which applying `fun` returned a truthy value, and
+  a second keyword list with all the elements for which applying `fun` returned
+  a falsy value (`false` or `nil`).
+
+  See the examples below.
```
```suggestion
```

ericmj commented on lib/elixir/lib/map.ex:
```diff
@@ -838,6 +838,53 @@ defmodule Map do
+    do_split_with(next, fun)
+  end
+
+  defp do_split_with(acc \\ {[], []}, next, fun)
```
Default arguments on private functions are usually avoided, pass the arguments directly instead.

ericmj commented on lib/elixir/lib/map.ex:
```diff
@@ -838,6 +838,53 @@ defmodule Map do
+    {:maps.from_list(while_true), :maps.from_list(while_false)}
+  end
+
+  defp do_split_with({while_true, while_false}, {key, value, iter}, fun) do
```
If we wrap in a tuple we need to allocate it for every iteration. Instead of the tuple use two arguments instead.

camilleryr replied:
That makes complete sense, I hadn't thought of that! PR updated to pass the accumulators separately. Thanks!

ericmj commented on lib/elixir/lib/map.ex:
```diff
@@ -838,6 +838,53 @@ defmodule Map do
+    do_split_with(next, fun)
+  end
+
+  defp do_split_with(acc \\ {[], []}, next, fun)
```
Default arguments on private functions are usually avoided, pass the arguments directly instead.

camilleryr replied:
Removed!

josevalim commented:
After solving today's Advent of Code, which I assume was the prompt for this function, I am no longer convinced this is the way to go.

My overall solution didn't use maps and the initial approach that used maps would return maps only then to immediately traverse them, which would be more efficiently done if they were kept as lists.

I am also planning to add notes to `Map.map+Map.filter` to note that, if you are doing multiple traversals, the Enum implementation is more efficient.

Thoughts?

camilleryr commented:
> After solving today's Advent of Code, which I assume was the prompt for this function, I am no longer convinced this is the way to go.
> 
> My overall solution didn't use maps and the initial approach that used maps would return maps only then to immediately traverse them, which would be more efficiently done if they were kept as lists.
> 
> I am also planning to add notes to `Map.map+Map.filter` to note that, if you are doing multiple traversals, the Enum implementation is more efficient.
> 
> Thoughts?

The desire for this definitely came out of one of the recent AoC puzzles!  

I will differ to you to weigh the value of this being added to the standard lib though!  My thought for the proposal was strictly that I looked for it, and it was not there!  And with the addition of Map.map and Map.filter it seemed like there was a whole where `split_with` could be.

I feel like there are valid use cases for this particular operation, but they me be fairly few and far between, and if `split_with` is actually the right tool for the job - you can easily replicate it with pieces for the existing standard lib.

I think the tough question then becomes do you not include a function that is more performant for some subset of potential use-cases because it may lead to less performant code because its composed with other operations in other use-cases?

I think you are in a much better place to answer that question then I am!  I'll trust your decision either way!

josevalim commented:
Thank you @camilleryr! For now, I am going with a no. We can always revisit once we have more use cases!

camilleryr commented:
@josevalim - I ran across a use case in the wild where Map.split_with would have been handy and just wanted to see if you thoughts about these apis had changed any sine last year (I know Map.map / Map.filter were added, deprecated, and then un deprecated - so with those in the standard library for good, didnt know if you would be interested in adding these in)

josevalim commented:
Given the undecisiveness around Map.filter/Map.map, I would prefer to not add more Enum-like functions for now.

josevalim commented:
Actually, I have checked the Enum module and it seems this is the only function that would also apply to maps. So I am reopening and reviewing. Meanwhile, can you also please add `MapSet.split_with`?

josevalim commented on lib/elixir/lib/keyword.ex:
```diff
@@ -1166,6 +1166,46 @@ defmodule Keyword do
+      {[], []}
+
+  """
+  @doc since: "1.14.0"
```
```suggestion
  @doc since: "1.15.0"
```

josevalim commented on lib/elixir/lib/keyword.ex:
```diff
@@ -1166,6 +1166,46 @@ defmodule Keyword do
+  @doc since: "1.14.0"
+  @spec split_with(t, ({key, value} -> as_boolean(term))) :: {t, t}
+  def split_with(keywords, fun) when is_list(keywords) and is_function(fun, 1) do
+    fun = fn key_vaue_pair, {while_true, while_false} ->
```
```suggestion
    fun = fn key_value_pair, {while_true, while_false} ->
```

camilleryr replied:
Whoops!

josevalim commented on lib/elixir/lib/map.ex:
```diff
@@ -838,6 +838,51 @@ defmodule Map do
+      {%{}, %{}}
+
+  """
+  @doc since: "1.14.0"
```
```suggestion
  @doc since: "1.15.0"
```

camilleryr commented:
> Actually, I have checked the Enum module and it seems this is the only function that would also apply to maps. So I am reopening and reviewing. Meanwhile, can you also please add `MapSet.split_with`?

I have updated the PR and added `MapSet.split_with`

josevalim commented on lib/elixir/lib/keyword.ex:
```diff
@@ -1185,6 +1185,46 @@ defmodule Keyword do
+  Splits the `keywords` into two keyword lists according to the given function
+  `fun`.
+
+  Splits the given `keywords` into two keyword lists by calling the provided
+  `fun` which receives each `{key, value}` pair in the `keywords` as its only
```
```suggestion
  The provided `fun` receives each `{key, value}` pair in the `keywords` as its only
```

josevalim commented on lib/elixir/lib/map.ex:
```diff
@@ -855,6 +855,51 @@ defmodule Map do
+  @doc """
+  Splits the `map` into two maps according to the given function `fun`.
+
+  Splits the given `map` into two maps by calling the provided `fun` which
+  receives each `{key, value}` pair in the `map` as its only argument. Returns
```
```suggestion
  `fun` receives each `{key, value}` pair in the `map` as its only argument. Returns
```

josevalim commented on lib/elixir/lib/map_set.ex:
```diff
@@ -375,6 +375,37 @@ defmodule MapSet do
+  @doc """
+  Splits the `MapSet` into two `MapSet`s according to the given function `fun`.
+
+  Splits the given `MapSet` into two maps by calling the provided `fun` which
+  receives each element in the `MapSet` as its only argument. Returns
```
```suggestion
  `fun` receives each element in the `MapSet` as its only argument. Returns
```

josevalim commented on lib/elixir/lib/map_set.ex:
```diff
@@ -375,6 +375,37 @@ defmodule MapSet do
+
+  ## Examples
+
+      iex> {while_true, while_false} = MapSet.split_with(MapSet.new([1, 2, 3, 4]), fn v -> rem(v, 2) == 0 end)
+      iex> while_true
+      #MapSet<[2, 4]>
+      iex> while_false
+      #MapSet<[1, 3]>
+
+      iex> {while_true, while_false} = MapSet.split_with(MapSet.new(), fn {_k, v} -> v > 50 end)
+      iex> while_true
+      #MapSet<[]>
+      iex> while_false
+      #MapSet<[]>
```
```suggestion
      iex> {while_true, while_false} = MapSet.split_with(MapSet.new([1, 2, 3, 4]), fn v -> rem(v, 2) == 0 end)
      iex> while_true
      MapSet.new([2, 4])
      iex> while_false
      MapSet.new([1, 3])

      iex> {while_true, while_false} = MapSet.split_with(MapSet.new(), fn {_k, v} -> v > 50 end)
      iex> while_true
      MapSet.new([])
      iex> while_false
      MapSet.new([])
```

camilleryr replied:
Ohhh! The new inspect implementation! I realized I was working on a really old branch and didnt see the difference after I rebased! Very Nice!

josevalim commented:
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
